Parsing

nj-wilson opened the pull request:
Adds ability to import text into hazel as a program, as well as exporting the current program to plain text.

Currently is part of the hamburger menu as separate options (could easily be combined).
Import errors are shown, but currently are quite rudimentary, only showing what character caused a problem.

A couple commits are still local, for cleanup and additional tests.
There is a behavior difference with regard to empty or comment lines, but nothing that impacts correct programs.

ExternalLineBreak was added to Layout.t, but may be better as an Annot (for a future PR).
The module naming for the parser and lexer are somewhat confusing, i.e. Parsing.Parser, and Lexing.Lexer. Could use better names.

nj-wilson committed:
Initial testing with Menhir (monolithic)

incremental

nj-wilson committed:
Switch to using UHExp types

nj-wilson committed:
Cleanup parsing directory

nj-wilson committed:
Simplification, working Plus and Minus

nj-wilson committed:
Add MULT and DIV, use inline op rule

nj-wilson committed:
Lexer: Use Keyword table

nj-wilson committed:
Parsing: Support Let bindings, paranthesized exprs, grammar changes

nj-wilson committed:
Parser: Simple lambda

nj-wilson committed:
Parsing: small fixes

nj-wilson committed:
Parser: Remove wrapper rules

nj-wilson committed:
Parser: Move logic into functions

nj-wilson committed:
Parser: Parenthesized patterns

nj-wilson committed:
Parser: Use functions instead of directly constructing UHExp/Pat

nj-wilson committed:
Parser: Simple function application

Should be looked at again, something such as `(3+1) 3` does parse in hazel, albeit with `3+1` being in a hole. This cannot parse that correctly.

nj-wilson committed:
Parser: Refactor toplevel to more fit how UHExp types are.

nj-wilson committed:
Parser: Fix function application helper

nj-wilson committed:
Parser: Add Comment and Empty lines

nj-wilson committed:
Parser: Add Case, Rules

nj-wilson committed:
Parser: Refactor grammar away from UHExp

Wasn't worth the trouble

nj-wilson committed:
Parser: Add Cons, Nil

nj-wilson committed:
Parser: Add type annotations

nj-wilson committed:
Parser: Bool lits

nj-wilson committed:
Parser: Add floats

nj-wilson committed:
WIP: Remove trivial helpers

nj-wilson committed:
WIP: Simplify rules, reduce mk_seq calls

nj-wilson committed:
Parser: Add injections

nj-wilson committed:
Lexer: Fix Empty lines with whitespace

nj-wilson committed:
Lexer: Fix identifiers

nj-wilson committed:
Parser: Add missing annotation types

nj-wilson committed:
Parser: Add pattern constants

nj-wilson committed:
Small fixes

multiple semicolon delimited lines, consolidate case rule

nj-wilson committed:
Add missing tuple handling

nj-wilson committed:
Moved parser into hazelcore

nj-wilson committed:
Layout: add new Memo/WeakMemo modules

Move memoize into the WeakMemo module.
Memo.DummyMemo for faking memoize in core.

nj-wilson committed:
UHDoc_[Exp,Pat,Typ]: functorize to take in a layout/Memo module, fix uses

nj-wilson committed:
Break out Settings type into DHSettings module

nj-wilson committed:
web/layout: Move necessary files into core/layout

nj-wilson committed:
Parser: Dune file cleanups

nj-wilson committed:
Lexer: Add Hazel characters

string_of_layout currently uses non-breaking spaces, as well as lambda characters. Handle these in lexer

nj-wilson committed:
Parser: Move logic into functions, simplify rule actions

Attempts to make grammar rules easier to read

nj-wilson committed:
Parser: rename expr/expr_ -> line/expr

nj-wilson committed:
WIP: Add Import functionality from parsing, cleanup Printing

nj-wilson committed:
Parser: type annotations in let lines

nj-wilson committed:
WIP: add external linebreaks

nj-wilson committed:
Fixes from rebasing onto dev

nj-wilson committed:
Parser: remove unneeded precedences

nj-wilson committed:
Move import/export into popups

nj-wilson committed:
Add misc files

Parse tests, printing

nj-wilson committed:
Remove Serialize button

Was moved to menu

nj-wilson committed:
remove 'hazel' from parser, lexer names

Should probably revisit, names are kinda messy. Perhaps change Parsing to Driver?

nj-wilson committed:
Print out empty holes as _?

nj-wilson committed:
Parser: add empty holes

nj-wilson committed:
WIP: import errors

nj-wilson committed:
Remove unneeded modules

nj-wilson committed:
slight externalLinebreak tweaks

nj-wilson committed:
Rework parser tests

cyrus- commented on src/hazelcore/parse/dune:
```diff
@@ -0,0 +1,14 @@
+(include_subdirs no)
```
Since it can be its own library, lets move the parser out to the `src` directory instead of it being inside `hazelcore`. You can call it `hazeltext`.

cyrus- commented on src/hazelcore/parse/print.re:
```diff
@@ -0,0 +1,19 @@
+    };
+  let (s, _) = go(0, 0, l);
+  s;
+};
```
let's add an end-to-end `print_exp : UHExp.t -> string` as well

cyrus- commented on src/hazelcore/parse/test/test.ml:
```diff
@@ -0,0 +1,75 @@
+    let lay = Pretty.LayoutOfDoc.layout_of_doc ~width:100 ~pos:0 doc in
+    let l = match lay with
+    | Some l -> l
+    | None -> Pretty.Layout.Text("")
```
this should be a failure case, not ""

cyrus- commented on src/hazelcore/parse/parse.mly:
```diff
@@ -0,0 +1,223 @@
+;
+
+pat:
+  pat COLONCOLON pat { mk_binop $1 Operators_Pat.Cons $3 }
```
type annotations can now appear on any pattern, not just at the top level of a let or lambda

cyrus- commented on src/hazelcore/parse/test/test.ml:
```diff
@@ -0,0 +1,75 @@
+          end
+        }
+      }
+  in map"
```
let's add a few tests of longer opseqs involving all of the various operators

cyrus- commented on src/pretty/Layout.re:
```diff
@@ -6,6 +6,7 @@ type t('annot) =
   | Text(string) // Invariant: contains no newlines. Text("") is identity for `Cat`
   | Cat(t('annot), t('annot)) // Associative
   | Linebreak
+  | ExternalLinebreak
```
Let's use Annot in this PR rather than changing pretty

nj-wilson committed:
Popup fixups

nj-wilson committed:
Move Parsing out to own library (hazeltext)

nj-wilson committed:
Raise Exception on No Layout for Print

nj-wilson committed:
Formatting

nj-wilson committed:
Move ExternalLinebreak to UHAnnot instead of pretty/layout

nj-wilson committed:
hazeltext/print: show holes as just underscore

nj-wilson committed:
hazeltext/print: just kidding, meant '?' for holes

nj-wilson committed:
hazeltext/print: convert to ocaml

nj-wilson committed:
lexer: missed changing hole to '?'

nj-wilson committed:
parser: allow typ annotations in any pattern

nj-wilson committed:
parser, lexer: properly handle float ops

nj-wilson committed:
parser: handle invalid text

nj-wilson committed:
Lexer: Redo Int and Float regexps

Modelled after ocaml definitions

nj-wilson committed:
Parser: Additional tests

nj-wilson committed:
hazeltext: handle wildcard

nj-wilson committed:
Parser: replace redundant mk_seq with Seq.wrap

nj-wilson committed:
hazeltext/parse: minor de-verbosing

nj-wilson committed:
Merge branch 'dev' into parsing

nj-wilson committed:
hazeltext/parse: wildcards in expression position should be invalid text

nj-wilson committed:
hazeltext/parse: Fix sequential typ annotations

nj-wilson committed:
WIP hazeltext/parse: temporarily remove whitespace with newline rule

nj-wilson committed:
hazeltext/parse: rework block, line

Split out expression lines into own rule, and redefine blocks properly as a list of lines

nj-wilson committed:
hazeltext: Fix renumbering holes in tests and Popup

nj-wilson committed:
Merge branch 'dev' into parsing

nj-wilson committed:
Merge branch 'dev' into parsing

nj-wilson committed:
UndoHistory: Support Import model actions

Add Import action_group

nj-wilson committed:
hazelweb: Remove Export popup

Consolidate into a single popup, 'load current program' handles the same
functionality.

nj-wilson committed:
hazeltext/parsing: rework ast_of_lexbuf to return Result instead of a pair

nj-wilson committed:
Hazelweb: Re-style import popup

nj-wilson committed:
Hazelweb: Rename Import Popup to Text Editor

nj-wilson committed:
Hazelweb/Popup: rework multiple text-box manipulation functions

Additionally fixes a bug with the text-area not updating after clicking
edit

nj-wilson committed:
Hazeltext: change ast_of_lexbuf to fix holes

nj-wilson committed:
wip: grammar rules

nj-wilson committed:
Hazeltext/parsing: add ast_of_string wrapper function

nj-wilson committed:
Hazeltext/parse: rearrange expr grammar rules

Fixes small things such as cases in function position

nj-wilson committed:
Hazeltext/parse: reorganize tokens

nj-wilson committed:
Hazeltext/parse: misc organization

nj-wilson committed:
Merge branch 'dev' into parsing

nj-wilson committed:
Hazeltext/parse: simply parts of grammar

nj-wilson committed:
Merge branch 'dev' into parsing

nj-wilson committed:
Hazeltext: Lambda -> Fun

nj-wilson committed:
rewrite Examples to be parsed text instead of AST structures

cyrus- committed:
added .ocamlformat file to suppress build warning

cyrus- reviewed (changes requested):
Given that the popup text editor looks almost exactly like the structure editor UI, what if we actually had it integrated into the main editor as an editing mode toggle -- you can either edit using the structure editor or text editor, and there was a toggle to go back and forth which would trigger parsing/pretty printing as necessary?

In text mode, we could have it attempt to parse on every edit and run the evaluator as usual, but if parsing fails then you get the parse error instead of the result?

nj-wilson commented:
> 

Is an interesting idea, I shall look into it.
My only questions would be on how to display errors resulting in trying to go back to structured, though maybe that'll be taking care of from running the evaluator during edit. Though I do wonder if that would be problematic to run on every character typed.

I'll also have to see how easy it would be to switch the main text box between modes, or do some other method.

nj-wilson committed:
add missing and, or operators

cyrus- commented:
You can use the same area in the UI that results are displayed to show errors, since no result is available when parsing fails, and the button to toggle back to structure editing can be disabled when there is a parse error. Don't worry about the performance cost of running the parser on each edit. If that becomes an issue down the line we can look into incremental parsing approaches, but nothing at the scale we're using Hazel should cause an issue. Learn OCaml already parses and pretty prints on every character. 

Shouldn't be too difficult GUI wise with the model-view-controller approach. When in text mode, render the text box you already have set up from the pop up in your branch, otherwise render the existing structure editor box.

cyrus- commented:
no longer relevant with the two other PRs.
